cpp/src/lists/interleave_columns.cu

```diff
@@ -0,0 +1,359 @@
+
+  // The array of int8_t to store validities for list elements.
+  auto validities = rmm::device_uvector<int8_t>(has_null_mask ? num_output_lists : 0, stream);
+
+  // Compute list sizes and validities.
+  thrust::transform(
+    rmm::exec_policy(stream),
+    thrust::make_counting_iterator<size_type>(0),
+    thrust::make_counting_iterator<size_type>(num_output_lists),
+    d_offsets,
+    [num_cols,
+     table_dv     = *table_dv_ptr,
+     d_validities = validities.begin(),
+     has_null_mask] __device__(size_type const dst_list_id) {
+      auto const src_col_id     = dst_list_id % num_cols;
+      auto const src_list_id    = dst_list_id / num_cols;
+      auto const& src_lists_col = table_dv.column(src_col_id);
+      if (has_null_mask) {
+        d_validities[dst_list_id] = static_cast<int8_t>(src_lists_col.is_valid(src_list_id));
+      }
+      auto const src_list_offsets =
+        src_lists_col.child(lists_column_view::offsets_column_index).data<size_type>() +
+        src_lists_col.offset();
+      return src_list_offsets[src_list_id + 1] - src_list_offsets[src_list_id];
+    });
+
+  // Compute offsets from sizes.
+  thrust::exclusive_scan(
+    rmm::exec_policy(stream), d_offsets, d_offsets + num_output_lists + 1, d_offsets);
+
+  return {std::move(list_offsets), std::move(validities)};
+}
+
+/**
+ * @brief Compute string sizes, string validities, and interleave string lists functor.
+ *
+ * This functor is executed twice. In the first pass, the sizes and validities of the output strings
+ * will be computed. In the second pass, this will interleave the lists of strings of the given
+ * table of lists columns.
+ */
+struct compute_string_sizes_and_interleave_lists_fn {
+  table_device_view const table_dv;
+
+  // Store list offsets of the output lists column.
+  offset_type const* const dst_list_offsets;
+
+  // Flag to specify whether to compute string validities.
+  bool const has_null_mask;
+
+  // Store offsets of the strings in the chars column.
+  offset_type* d_offsets{nullptr};
+
+  // If d_chars == nullptr: only compute sizes and validities of the output strings.
+  // If d_chars != nullptr: only concatenate strings.
+  char* d_chars{nullptr};
+
+  // We need to set `1` or `0` for the validities of the strings in the child column.
+  int8_t* d_validities{nullptr};
+
+  __device__ void operator()(size_type const dst_list_id)
+  {
+    auto const num_cols    = table_dv.num_columns();
+    auto const src_col_id  = dst_list_id % num_cols;
+    auto const src_list_id = dst_list_id / num_cols;
+
+    auto const& src_lists_col = table_dv.column(src_col_id);
+    if (has_null_mask and src_lists_col.is_null(src_list_id)) { return; }
+
+    auto const src_list_offsets =
+      src_lists_col.child(lists_column_view::offsets_column_index).data<size_type>() +
+      src_lists_col.offset();
+    auto const& src_child = src_lists_col.child(lists_column_view::child_column_index);
+    auto const src_child_offsets =
```

Comment:
You might consider calling this  "strings" and "string_offsets".  Using "src" here obfuscates things a little bit.